### ISSUE TITLE
[pgbench] pin_node is NOT optional in ripsaw_v1alpha1_pgbench_cr.yaml

### ISSUE THREAD

**Issue by netzzer** [cloud-bulldozer/ripsaw#144]
Leaving pin-node blank caused the client pod to not create. Adding a value such as `pin_node: "ip-10-0-134-143"` in ripsaw_v1alpha1_pgbench_cr.yaml fixed the problem and the test ran as expected after this change.

```
$ oc get benchmark
NAME                TYPE      AGE
pgbench-benchmark   pgbench   56s
[ec2-user@ip-172-31-14-128 ripsaw-dustin]$ oc describe benchmark pgbench-benchmark
Name:         pgbench-benchmark
Namespace:    ripsaw
Labels:       <none>
Annotations:  <none>
API Version:  ripsaw.cloudbulldozer.io/v1alpha1
Kind:         Benchmark
Metadata:
  Creation Timestamp:  2019-07-08T23:39:49Z
  Generation:          1
  Resource Version:    91783
  Self Link:           /apis/ripsaw.cloudbulldozer.io/v1alpha1/namespaces/ripsaw/benchmarks/pgbench-benchmark
  UID:                 ad4debce-a1d9-11e9-acf0-0268146ce15c
Spec:
  Workload:
    Args:
      Clients:
        4
        8
      cmd_flags:  
      Databases:
        db_name:       sampledb
        Host:          172.30.86.214
        Password:      wTRfg5vxpmtfkYKA
        pin_node:      <nil>
        Port:          <nil>
        User:          user8LH
      init_cmd_flags:  
      run_time:        300
      Samples:         2
      scaling_factor:  30
      Threads:         4
      Timeout:         5
      Transactions:    <nil>
    Name:              pgbench
Status:
  Conditions:
    Last Transition Time:  2019-07-08T23:40:53Z
    Message:               Running reconciliation
    Reason:                Running
    Status:                False
    Type:                  Running
    Ansible Result:
      Changed:             2
      Completion:          2019-07-08T23:40:56.93001
      Failures:            1
      Ok:                  5
      Skipped:             6
    Last Transition Time:  2019-07-08T23:40:57Z
    Message:               An unhandled exception occurred while running the lookup plugin 'template'. Error was a <class 'ansible.errors.AnsibleError'>, original message: Unexpected templating type error occurred on (---
kind: Job
apiVersion: batch/v1
metadata:
  name: '{{ meta.name }}-pgbench-client-{{ item.0|int + 1 }}'
  namespace: '{{ operator_namespace }}'
spec:
  ttlSecondsAfterFinished: 600
  template:
    metadata:
      labels:
        app: pgbench-client
    spec:
      containers:
      - name: benchmark
        image: "quay.io/cloud-bulldozer/pgbench:latest"
        command: ["/bin/sh", "-c"]
        args:
          - "export PGPASSWORD='{{ item.1.password }}';
             export pgbench_auth='-h {{ item.1.host }} -p {% if item.1.port is defined and item.1.port|int > 0 %} {{ item.1.port }} {% else %} {{ db_port }} {% endif %} -U {{ item.1.user }}';
             echo 'Init Database {{ item.1.host }}/{{ item.1.db_name }}';
             pgbench $pgbench_auth -i -s {{ pgbench.scaling_factor }} {{ pgbench.init_cmd_flags }} {{ item.1.db_name }};
             if [ $? -eq 0 ]; then
               echo 'Waiting for start signal...';
               redis-cli -h {{ bo.resources[0].status.podIP }} lpush pgb_client_ready {{ item.0|int }};
               while true; do
                 if [[ $(redis-cli -h {{ bo.resources[0].status.podIP }} get pgb_start) =~ 'true' ]]; then
                   echo 'GO!';
                 {% for clients in pgbench.clients %}
                   echo '';
                   echo 'Running PGBench with {{ clients }} clients on database {{ item.1.host }}/{{ item.1.db_name }}';
                   for i in `seq 1 {{ pgbench.samples|int }}`; do
                     echo \"Begin test sample $i of {{ pgbench.samples }}...\";
                     pgbench $pgbench_auth -c {{ clients }} -j {{ pgbench.threads }} {% if pgbench.transactions is defined and pgbench.transactions|int > 0 %} -t {{ pgbench.transactions }} {% elif pgbench.run_time is defined and pgbench.run_time|int > 0 %} -T {{ pgbench.run_time }} {% endif %} -s {{ pgbench.scaling_factor }} {{ pgbench.cmd_flags }} {{ item.1.db_name }};
                   done;
                 {% endfor %}
                 else
                   continue;
                 fi;
                 break;
               done;
             fi"
      restartPolicy: OnFailure
{% if item.1.pin_node is defined and item.1.pin_node|length and item.1.pin_node is not sameas "" %}
      nodeSelector:
        kubernetes.io/hostname: '{{ item.1.pin_node }}'
{% endif %}
): object of type 'NoneType' has no len()
    Reason:  Failed
    Status:  True
    Type:    Failure
Events:      <none>
```

**Comment by jtaleric** [comment]
Did you try with `pin_node` removed from the CR?

```
          # pin_node is an optional kubernetes hostname to which the pgbench pod will be pinned
          pin_node: 
```

When looking at the code. 

```

{% if item.1.pin_node is defined and item.1.pin_node|length and item.1.pin_node is not sameas "" %}
      nodeSelector:
        kubernetes.io/hostname: '{{ item.1.pin_node }}'
{% endif %}

```

We might want to check the length of pin_node against a value, for example : 

```
{% if item.1.pin_node is defined and item.1.pin_node|length > 1 and item.1.pin_node is not sameas "" %}
      nodeSelector:
        kubernetes.io/hostname: '{{ item.1.pin_node }}'
{% endif %}
```

**Comment by dustinblack** [comment]
Tested these conditions:

1. `pin_node` not defined: **passed**
2. `pin_node` defined with a valid node name (from oc get nodes): **passed**
3. `pin_node` defined but empty: **failed**
4. `pin_node` defined as '': **passed**
5. `pin_node` defined as false: **failed**

5 above was just an extra test, so I'm sure 3 is what got you. 

In the case of 3, we get the error:
```
object of type 'NoneType' has no len()
```
So I don't think there will be a difference in checking `item.1.pin_node|length` vs `item.1.pin_node|length > 0` since it's simply angry that you can't check length on a `NoneType` variable (defined but without a value).

I'll look at this again tomorrow after some sleep and coffee. I'm sure there is a simple solution in the if statement that I'm not seeing right now.

**Comment by jtaleric** [comment]
Instead of us having to duplicate logic (like we currently are), how about we create `role/pretasks` that runs before all the other roles. 

The responsibility of this role is to check for things like this in the CR, reducing the duplicate checks we currently have to do. 

In this case we would need 

```
item.1.ping_node is not none
```

I don't think this will eliminate all checks in other roles, but it will reduce things, additionally catch the error early... @dustinblack @aakarshg @dry923 thoughts?

**Comment by dustinblack** [comment]
@jtaleric So this might for instance convert any defined but empty variables from the CR into blank values ('') instead, so that the validation logic within the main role is simplified?

**Comment by dustinblack** [comment]
Tested with simplified `if` logic:

```
{% if item.1.pin_node is defined and item.1.pin_node %}
```
And the results look good:

1. `pin_node` not defined: **passed**
2. `pin_node` defined with a valid node name (from oc get nodes): **passed**
3. `pin_node` defined but empty: **passed** << winning
4. `pin_node` defined as '': **passed**
5. `pin_node` defined as false: **passed** << winning

**Comment by dustinblack** [comment]
Fixed with PR #145
